Add `amount::Display` - make formatting configurable

Kixunil opened the pull request:
This significatnly refactors the formatting code to make formatting more
configurable. The main addition is the `Display` type which is a
builder that can configure denomination or other things (possibly more
in the future).

Further, this makes all representations of numbers minimal by default,
so should be documented as a possibly-breaking change.

Because of the effort to support all other `fmt::Formatter` options this
required practically complete rewrite of `fmt_satoshi_in`. As a
byproduct I took the opportunity of removing one allocation from there.

Closes #709

dr-orlovsky commented:
the way `Amount` is displayed (and also parsed from string) is clearly poor. But I haven't got from the original discussion why just using formatting directives is not enough and we need to introduce dedicated `Display` type and builder.

Kixunil commented:
Because there's no way to pass more than two possible custom formats into `fmt::Formatter` (not abusing other things which we actually want to use for what they were designed for).

dr-orlovsky commented:
Concept ACK

apoelstra commented:
concept ACK. It's not clear from skimming the PR what happens when you just try to `Display` a bare amount, but I guess it should be the same as using `display_dynamic`?

Kixunil commented:
@apoelstra I didn't change the original implementation which was displaying amount in BTC with BTC suffix.

apoelstra commented:
Ok, that's a reasonable default too.

Kixunil commented:
Rebased, added tests, refactored&implemented `SignedAmount`

I'm not sure if the commits should be squashed though. Leaning a bit towards yes. What do you think?

Kixunil commented on src/util/amount.rs:
```diff
@@ -1459,18 +1849,22 @@ mod tests {
+        assert_eq!("0.50", format!("{:.2}", Amount::from_sat(50).display_in(D::Bit)));
+        //assert_eq!("-0.50", format!("{:.2}", SignedAmount::from_sat(-50).display_in(D::Bit)));
+        assert_eq!("0.10000000", format!("{:.8}", Amount::from_sat(100_000_00).display_in(D::Bitcoin)));
+        //assert_eq!("-100.00", format!("{:.2}", SignedAmount::from_sat(-10_000).display_in(D::Bit)));
```
Forgot to uncomment these, will do after deciding to squash/not squash.

thomaseizinger commented on src/util/amount.rs:
```diff
@@ -605,6 +757,62 @@ impl ::core::iter::Sum for Amount {
 
+/// A helper/builder that displays amount with specified settings.
+///
+/// This provides richer interface than `fmt::Fromatter`:
```
```suggestion
/// This provides richer interface than `fmt::Formatter`:
```

thomaseizinger commented on src/util/amount.rs:
```diff
@@ -605,6 +757,62 @@ impl ::core::iter::Sum for Amount {
+///
+/// However this can still be combined with `fmt::Formatter` options to precisely control zeros,
+/// padding, alignment... The formatting works like floats from `core` but note that precision will
+/// be **never** lossy - that means no rounding.
```
```suggestion
/// **never** be lossy - that means no rounding.
```

thomaseizinger commented on src/util/amount.rs:
```diff
@@ -605,6 +757,62 @@ impl ::core::iter::Sum for Amount {
+/// padding, alignment... The formatting works like floats from `core` but note that precision will
+/// be **never** lossy - that means no rounding.
+///
+/// See [`Amount::display_in`] and [`Amount::display_dynamic`] to construct this.
```
```suggestion
/// See [`Amount::display_in`] and [`Amount::display_dynamic`] on how to construct this.
```

thomaseizinger commented on src/util/amount.rs:
```diff
@@ -605,6 +757,62 @@ impl ::core::iter::Sum for Amount {
+impl Display {
+    /// Makes subsequent calls to `Display::fmt` display denomination.
+    pub fn show_denomination(mut self) -> Self {
+        match &mut self.style {
+            DisplayStyle::FixedDenomination { show_denomination, .. } => *show_denomination = true,
+            // No-op because dynamic denomination is always shown
+            DisplayStyle::DynamicDenomination => (),
+        }
+        self
```
```suggestion
	Self {
		style: match self.style {
		    DisplayStyle::FixedDenomination { denomination, .. } => DisplayStyle::FixedDenomination { show_denomination: true, denomination },
	            DisplayStyle::DynamicDenomination => DisplayStyle::DynamicDenomination,
		},
		..self
	}
```

Note: Untested for MSRV compatibility but this could be the non-mutable style of achieving the same thing (I think).

thomaseizinger replied:
Yeah no drama :)

thomaseizinger reviewed (commented):
concept ACK

Some doc nits.

Kixunil commented:
The MSRV is soooo old :sob:

tcharding commented on src/util/amount.rs:
```diff
@@ -605,6 +757,62 @@ impl ::core::iter::Sum for Amount {
+///
+/// * Ability to select denomination
+/// * Show or hide denomination
+/// * Dynamically-selected denomination - show in sats if less than 1 BTC. 
```
This line has trailing whitespace.

tcharding reviewed (commented):
Wow, thorough testing, nice effort.

tcharding commented on src/util/amount.rs:
```diff
@@ -605,6 +759,62 @@ impl ::core::iter::Sum for Amount {
+
 /// SignedAmount
 ///
 /// The [SignedAmount] type can be used to express Bitcoin amounts that supports
```
There are two instances of this typo.

Kixunil replied:
How about having an issue where people can link typos and once there are many we do a larger batch of fixes?

tcharding replied:
That's a good idea, should save time.

Kixunil replied:
#828

dr-orlovsky commented on src/util/amount.rs:
```diff
@@ -717,15 +927,40 @@ impl SignedAmount {
+    /// Returns the absolute value as satoshis.
+    ///
+    /// This is the implementation of `unsigned_abs()` copied from `core` to support older MSRV.
+    fn as_sat_abs(self) -> u64 {
```
nit: I think according to our latest naming discussions this should be `to_sat_abs`

tcharding replied:
I nominate `to_`/`as_`/`into_` for category 'most annoying function naming convention to get correct' and apologize profusely for opening this particular can of worms.

thomaseizinger replied:
Clippy makes it reasonably easy though: https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

tcharding replied:
Oh sweeeet - I did not know about that lint. Cheers!

Kixunil replied:
Oh, yes, I made the name in line with the other methods without thinking.

dr-orlovsky reviewed (approved):
utACK 31ae76061b46c37023595f39d27505cbda6e4042

dr-orlovsky reviewed (approved):
ACK 4f1200d629d6dbbfae76d0fdddf4eac297ee989f

tcharding reviewed (approved):
LGTM

sanket1729 reviewed (approved):
ACK 4f1200d629d6dbbfae76d0fdddf4eac297ee989f
Thanks for being super careful with this and the elaborate test cases. I learned a lot about format options reviewing this. Hopefully, this settles the Amount Display issue for a long time.  

We should add this in release notes as a breaking change

tcharding commented:
ACK 4f1200d629d6dbbfae76d0fdddf4eac297ee989f
